### PR TITLE
first implementation

### PR DESCRIPTION
**Pull request by 31415us**
i had some issues naming stuff, since i don't know the jargon very well .-P

there currently is a possible data race when updating module parameters while doing computation, for simplicity i didn't handle it yet, let me know if i should put it in a critical section

generally i wasn't very sure if this is want you want, please let me know if you need additional functionality

### CODE REVIEW

File: src/trajectory_tracking.c [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+/* public module prototypes */
+uint8_t tracy_linear_controller(const tracking_error_t * current_error,
+                                const reference_velocity_t * reference_velocity,
+                                feedback_rule_t * output);
+
+uint8_t tracy_set_controller_params(float fancy_greek_c, float g);
+
+/* module variables */
+
+static float param_fancy_greek_c = DEFAULT_PARAM_FANCY_GREEK_C;
```
Comment (antoinealb) [inline src/trajectory_tracking.c]: The letter is Zeta ;) But I think "damping coefficient" would be more descriptive.

File: src/trajectory_tracking.h [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+/**
+ * compute the linear controller output
+ *
+ * @param current_error error between where you want to be and where you are 
+ *                      at this timestep
+ * @param reference_velocity velocity you want to be at currently
+ * @param output pointer to memory to store result
+ *
+ * @return 0 if everything went fine, or something else on failure
```
Comment (Stapelzeiger) [inline src/trajectory_tracking.h]: You don't have to check for NULL pointers. The problem is, that if the function can fail we have to handle the error case in the caller. It is a lot easier to have math functions that never fail

File: src/trajectory_tracking.h [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    float theta_error;
+} tracking_error_t;
+
+/**
+ * velocity of reference in polar robot coordinates
+ *
+ */
+typedef struct {
+    float radial_velocity;
+    float angular_velocity;
```
Comment (antoinealb) [inline src/trajectory_tracking.h:26]: "tangential_velocity" instead of radial velocity. At least I think.

Comment (31415us) [inline src/trajectory_tracking.h:26]: i put radial_velocity for consistency with the reference paper but will change it^^ tangential_velocity makes more sense

File: src/trajectory_tracking.h (hunk 2 above)
Comment (31415us) [inline src/trajectory_tracking.h]: ok so i'd change the return type to '''void''' and fail hard on null pointer dereference? (=

File: src/trajectory_tracking.c (hunk 1 above)
Comment (31415us) [inline src/trajectory_tracking.c]: haha ok thanks (=

**Comment by antoinealb** [comment]
Also quite important question what are rule1 and rule2 ? I think we need to chat a bit on Wednesday :smiley_cat:

**Comment by 31415us** [comment]
it's the feedback part of the velocity command vector u but yeah i agree that i didn't really understand much^^

File: src/trajectory_tracking.h (hunk 3 above)
Comment (SyrianSpock) [inline src/trajectory_tracking.h:26]: In the paper, it's tangential velocity, it's fundamentally different than radial velocity (perpendicular by definition) :)

**Comment by 31415us** [comment]
TODO:
add feedforward path to controller
rename feedback_rule_t to  velocity_command_t (other ideas?)
rename fields of velocity_command_t (they represent commands for tangential and angular velocity of the robot)
add paper to repo

**Comment by 31415us** [comment]
addendum:
replace sqrt by sqrtf

**Comment by 31415us** [comment]
ready for re-review (=

**Comment by antoinealb** [comment]
LGTM.

**Comment by antoinealb** [comment]
Juste a change. We have two different velocity types which are identical... Should wemerge them as one ?

File: src/trajectory_tracking.h [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+#define DEFAULT_PARAM_G (60.0f)
+
+/**
+ * output of the linear controller
+ *
+ */
+typedef struct {
+    float tangential_velocity;
+    float angular_velocity;
+} velocity_command_t;
```
Comment (Stapelzeiger) [inline src/trajectory_tracking.h]: I suggest to define a struct instead of a typedef, since we don't provide getter functions.

**Comment by 31415us** [comment]
@antoinealb done
@Stapelzeiger done

**Pull request merged**